### PR TITLE
RFC: Newbie Code Examples directory and files

### PR DESCRIPTION
**Pull request by jlp765**
Add xdoc directory
Add readme.txt to say what the directory is for (copied/modified from the docs directory)
Add examples.rst as the top-level file
Add two code example (memfiles.rst includes a file from the test suite)

------
``example.rst`` is aimed at being a web page that can be refered to from other pages, and lists all the examples in the xdoc directory

There are two "types" of examples I've added:

1. includes the code as rst formated code example (``file_read_text.rst``)
2. includes an existing file from the test-suite into the rst file (so that if the original file changes, the changes are reflected in the rst file) (``memfiles.rst``)

### CODE REVIEW

File: xdoc/file_read_text.rst [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+* Provide some suitably sized (text) file called junkdata.txt when trying this example
+* compile with -d:release (the file reading is slower otherwise)
+  
+
+.. code-block:: nim
+  
+  import strutils, times, os, parsecsv, streams, memfiles
+  
+  let
+    fname = "junkdata.txt"
```
Comment (Araq) [inline xdoc/file_read_text.rst]: Use ``const`` here, examples need to be as idiomatic as possible.

File: xdoc/file_read_text.rst [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+.. code-block:: nim
+  
+  import strutils, times, os, parsecsv, streams, memfiles
+  
+  let
+    fname = "junkdata.txt"  
+
+  const
+    FGETSZ = 2500
```
Comment (Araq) [inline xdoc/file_read_text.rst]: Meh, I still don't see the point in ALLCAPS constants. So yes, it's a constant, why shout that fact? ;-)

File: xdoc/file_read_text.rst [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  var
+    t0, t1 = cpuTime()
+    F: File
+    line = ""
+    cntL = 0
+    cntC = 0
+  
+  var fgetStr: cstring = newString(FGETSZ+1)
+  
+  proc fgets(c: cstring, n: cint, f: File): cstring  {.importc: "fgets", header: "<stdio.h>".}
```
Comment (Araq) [inline xdoc/file_read_text.rst]: At least add a warning here that ``fgets`` is really a horrible thing and one should avoid it at all costs.

File: xdoc/file_read_text.rst [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  iterator fgetsLine(f: File): string =
+    while not isNil(fgets(fgetStr, FGETSZ, f)):
+      yield $fgetStr
+  
+  proc fgetsAll(f: File): string =
+    result = ""
+    for x in f.fgetsLine():
+      result.add(x)
+    
+  proc prepare() =
```
Comment (Araq) [inline xdoc/file_read_text.rst]: prepare+finalise is not idiomatic Nim. Idiomatic Nim uses a template that takes a ``body: untyped``.

File: xdoc/file_read_text.rst [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  proc finalise(msg: string) =
+    t1 = cpuTime()
+    let s = (msg & repeat(' ',25))[0..25]
+    echo s,($(t1-t0) & "00")[0..4], "  Lines: ", cntL, " chars: ",cntC
+  
+  # -- MAIN ---
+  proc main() =
+    # Check file exists
+    if not open(F, fname):
+      echo "Unable to find ", fname
```
Comment (Araq) [inline xdoc/file_read_text.rst]: quit("Unable to find " & fname)

File: xdoc/file_read_text.rst [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  
+  let
+    fname = "junkdata.txt"  
+
+  const
+    FGETSZ = 2500    
+    
+  var
+    t0, t1 = cpuTime()
+    F: File
```
Comment (Araq) [inline xdoc/file_read_text.rst]: 'F' is against NEP-1. Use 'f' or 'file' or something else instead.

File: xdoc/file_read_text.rst [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    # This avoids manipulating the file on a per-line basis
+    # and is probably not a practical example for most cases,
+    # but does highlight that you may make performance gains if
+    # you optimise your code for a specific situation
+    #
+    prepare()
+    if open(F, fname):
+      let x = F.readAll
+      cntL = x.countLines
+      cntC = x.len() - (cntL * "\n".len)
```
Comment (Araq) [inline xdoc/file_read_text.rst]: This is wrong. ``countLines`` can deal with any line ending (or should be able to deal with any line ending) and your "\n".len is platform specific, but the file might not use the platform's line endings... This is tricky.

**Comment by dom96** [comment]
I'm sorry but I don't like this. We should focus on short and to the point examples, like the ones in the ``db_*`` modules which I added recently(ish): http://nim-lang.org/docs/db_sqlite.html#examples-opening-a-connection-to-a-database. Nobody is going to read a wall of code to find out how a file can be read.

**Comment by Araq** [comment]
I really like it, but it should be the beginning of the "Nim cook book" that we'll officially have on the website and announce it. So, in this case, don't include these files from the stdlib, but instead link to the cook book. :-)

**Comment by jlp765** [comment]
It is obvious from your feedback @Araq that examples exemplifying the correct coding style is a good thing for newbies :-)

Definitely a "Cook book" rather than documenting a library.
I see it as a supplement/complement to module doco that the doco can link to for more extensive examples (not that the doco doesn't show basic code).

The module doco is proc based, where the "cook book" shows the flow of procs to accomplish basic functionality (making it easier to work out how to glue the procs together to povide some functionality).   It is aimed at newbies, not experienced people.

(eg, so I just learnt how to read a text file from system module, 
but how do I do the same thing using parsecsv, memfiles, ...
This allows newbies to come up to speed quicker)

**Pull request closed without merging**